tools/cldr-apps/src/org/unicode/cldr/web/SurveyForumParticipation.java

```diff
@@ -138,6 +132,133 @@ private int queryForumDb(Cell cell, String loc) throws SQLException {
     }
 
+    /**
+     * "Forum: Needing action (Requests and Discussions my organization has not responded to)"
+     *
+     * "Needs action" means:
+     *    - (Open request from other org AND (my org haven’t agreed or declined))
+     *    - OR (open discuss-only thread AND the last poster is not in my org)
+     *
+     * Compare JavaScript "passIfNeedingAction". Different on server since not "I" but "We" = "anybody in my org".
+     *
+     * @param conn
+     * @param loc
+     * @return the count
+     * @throws SQLException
+     *
+     * TODO: make this faster; it's slow
+     */
+    private int queryForumDbAct(String loc) throws SQLException {
+        PreparedStatement ps = null;
+        Integer count = -1;
+        try {
+            ps = prepareForumActQuery(loc);
+            ResultSet rs = ps.executeQuery();
+            count = 0;
+            while (rs.next()) {
+                int forumId = rs.getInt(1);
+                int postType =rs.getInt(2);
+                /*
+                 * For Discuss: look at the LAST reply in the thread, and skip the item if our org was its poster
+                 * For Request: look at each reply in the thread, and skip the item if our org agreed or declined
+                 */
+                if (postType == typeDiscuss) {
+                    if (!ourOrgPostedLast(forumId)) {
+                        ++count;
+                    }
+                } else { // Request
+                    if (!ourOrgAgreedOrDeclined(forumId)) {
+                        ++count;
+                    }
+                }
+            }
+        } finally {
+            DBUtils.close(ps);
+        }
+        return count;
+    }
+
+    private PreparedStatement prepareForumActQuery(String loc) throws SQLException {
+        String sql = "SELECT " + forumTable + ".id, " +  forumTable + ".type"
+            + " FROM " + forumTable
+            + " JOIN " + userTable
+            + " ON " + forumTable + ".poster=" + userTable + ".id"
+            + " WHERE " + forumTable + ".parent=-1"
+            + " AND " + forumTable + ".open=true"
+            + " AND " + forumTable + ".loc=?"
+            + " AND (" + forumTable + ".type=?"
+            + " OR NOT " + userTable + ".org=?)";
+
+        PreparedStatement ps = DBUtils.prepareForwardReadOnly(conn, sql);
+        ps.setString(1, loc);
+        ps.setInt(2, typeDiscuss);
+        ps.setString(3, org);
+        return ps;
+    }
+
+    /**
+     * Does the poster who made the last post in this thread belong to this org?
+     *
+     * @param conn the Connection
+     * @param forumRootId the id of the root (initial) forum post of the thread
+     * @return true or false
+     * @throws SQLException
+     */
+    private boolean ourOrgPostedLast(int forumRootId) throws SQLException {
+        boolean result = false;
+        PreparedStatement ps = null;
+        String sql = "SELECT " + userTable + ".org "
+            + " FROM " + forumTable
```

Comment:
Ideally these would be `?` params. But ok

Reply:
I'm not sure what benefit that would have. The table names don't vary for these queries. Would there be a performance benefit in spite of that?

Reply:
@btangmu sorry, correct these are table names, not parameters.